add multi-src: GalleryAdults

cuong-tran opened the pull request:
- New multi-src GalleryAdults:
   - Query for most popular tags list (first 3 pages) instead of hard-code
   - Support site's user favorite (login)
   - Support sorting (popular/latest) for tag browsing/search results
   - Setting to use full title/short title
   - Support popular list for language-specific
   - Fix author/group, add characters to description
   - Filter other languages out of search results
   - Support query search along with selected tags instead of ignore it
   - Support advanced include/exclude search with artist, group, character, parody, tag
- Migrate HentaiFox, AsmHentai, IMHentai to this multi-src
- Move HentaiFox from "en" to "all", add Japanese, Chinese & Korean (there are other languages too but too little number of titles)

Close #1792
Close #1966
Close #2515

Checklist:

- [X] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [X] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [X] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [X] Have not changed source names
- [X] Have explicitly kept the `id` if a source's name or language were changed
- [X] Have tested the modifications by compiling and running the extension through Android Studio
- [ ] Have removed `web_hi_res_512.png` when adding a new extension

cuong-tran committed:
Convert AsmHentai into multisrc GalleryAdults

Also convert some selector into function

cuong-tran committed:
Move HentaiFox to theme GalleryAdults

cuong-tran committed:
GalleryAdults: Fix search

cuong-tran committed:
MangaFox: fix TagFilter

cuong-tran committed:
fast page load & preference for slowly parsing image’s URL

cuong-tran committed:
AsmHentai: change shortTitle reference from list to swith

cuong-tran committed:
HentaiFox: add Korean

cuong-tran committed:
move HentaiFox from en to all

cuong-tran committed:
fix build

cuong-tran committed:
fix search: convert space to +

cuong-tran committed:
Request for tags list from site

cuong-tran committed:
Fix request for user’s favorites

cuong-tran committed:
- Optimize popular/latest request

- Improve ‘page’ param
- AsmHentai: support Latest/Popular

cuong-tran committed:
add SortFilter

cuong-tran committed:
Support multiple tags filter

cuong-tran committed:
Support exact match query

cuong-tran committed:
getTime

cuong-tran committed:
Fix Lang when searching

cuong-tran committed:
fix searchById

cuong-tran committed:
add language companion

cuong-tran committed:
Fix URL action

cuong-tran committed:
renovate

cuong-tran committed:
Support parsing json for page list

Fix generating page if less than 10 pages
HentaiFox: Random server selection

cuong-tran committed:
Migrate IMHentai to GalleryAdults

cuong-tran committed:
Preferences to support all methods for page querying

cuong-tran committed:
IMHentai: tagList

cuong-tran committed:
Expose some filters to child class, add more space to description

cuong-tran committed:
Fix Factory lang

cuong-tran committed:
Support browsing tags, speechless & favorite

cuong-tran committed:
IMHentai:

- support favorite browsing (require login)
- tag filter with queried popular tags
- advanced search for artist, group, character, parody, tag (include/exclude)
- remove language filters
- Fix language search

cuong-tran committed:
Move advance search to multi-src

cuong-tran committed:
Fix: hide speechless when not supported

cuong-tran committed:
add Hint to use comma

cuong-tran committed:
split code to Filters & Utils

cuong-tran committed:
bump version all 3 extensions

cuong-tran committed:
fix getTime

cuong-tran committed:
fix lint

cuong-tran committed:
Fix alternative name

cuong-tran committed:
improve cleanTag

AwkwardPeak7 commented on lib-multisrc/galleryadults/src/eu/kanade/tachiyomi/multisrc/galleryadults/GalleryAdults.kt:
```diff
@@ -0,0 +1,771 @@
+        get() = getString(PREF_PARSE_IMAGES, PARSE_METHOD_DEFAULT_VALUE)!!
+
+    override fun setupPreferenceScreen(screen: PreferenceScreen) {
+        ListPreference(screen.context).apply {
```
I don't think this should be an option for the user, enable the one which would work best for that perticular site in the code

cuong-tran replied:
I rewrote it so that it will fall back to later methods if one is not available. Beside from parsing JSON, all other methods would work (theoretically). It's not like user has to choose.

AwkwardPeak7 commented on lib-multisrc/galleryadults/src/eu/kanade/tachiyomi/multisrc/galleryadults/GalleryAdults.kt:
```diff
@@ -0,0 +1,771 @@
+
+    override val client: OkHttpClient by lazy {
+        network.cloudflareClient.newBuilder()
+            .setRandomUserAgent(
```
I don't think providing randomUA in a multisrc is needed, besides it doesn't really work great

AwkwardPeak7 commented on lib-multisrc/galleryadults/src/eu/kanade/tachiyomi/multisrc/galleryadults/GalleryAdults.kt:
```diff
@@ -0,0 +1,771 @@
+                            keys.addAll(
+                                filter.state.trim()
+                                    // any space except after a comma (we're going to replace spaces only between words)
+                                    .replace(Regex("""(?<!,)\s+"""), "+")
```
pull out these regex objects so they are not initialized every time in the loop

AwkwardPeak7 commented on lib-multisrc/galleryadults/src/eu/kanade/tachiyomi/multisrc/galleryadults/GalleryAdults.kt:
```diff
@@ -0,0 +1,771 @@
+                    )
+                }
+                .let { unfiltered ->
+                    if (mangaLang.isNotBlank()) unfiltered.filter { it.lang == mangaLang } else unfiltered
```
this type of filtering can generally cause issues. say a single page result only contained the language which is filtered out, leaving no results. then app will stop fetching next pages and instead show "No results found"

cuong-tran replied:
It's true. AsmHentail is doing this way.
HentaiFox is doing nothing, which includes all other languages' titles in search result. Maybe we can add a switch in filter list to let users decide for themselves?

AwkwardPeak7 replied:
sure that works

cuong-tran replied:
instead of user has to switch to all languages, I include 1 (and only 1) entry of other languages so it can still load next page

AwkwardPeak7 commented on lib-multisrc/galleryadults/src/eu/kanade/tachiyomi/multisrc/galleryadults/GalleryAdultsUtils.kt:
```diff
@@ -0,0 +1,134 @@
+import java.text.SimpleDateFormat
+import java.util.Calendar
+
+object GalleryAdultsUtils {
```
is this object namespace needed, you can simply add them to the main class

AwkwardPeak7 reviewed (changes requested):
is this a port of `GalleryAdults` parser from Kotatsu 🤔

~what about the other site in there~

cuong-tran committed:
move out of Object

cuong-tran committed:
move Regex out

cuong-tran committed:
remove RandomUA

cuong-tran committed:
fix build

cuong-tran committed:
remove images parsing setting, pick a default one

cuong-tran committed:
fix build

AwkwardPeak7 commented on lib-multisrc/galleryadults/res/mipmap-hdpi/ic_launcher.png:
not sure if this multisrc's default icon should be nhentai logo

cuong-tran replied:
this theme is indeed pretty much similar to nHentai with a few differences. Kotatsu also created nHentai using same multi-src. I'm thinking about also migrating nHentai to this theme but it will require a few differences.

vetleledaal replied:
If you're adding nhentai, you could also look into #2126, but that could just as easily be left for another PR

cuong-tran replied:
Sure! nHentai is a bit different so I also thinking will add it in another PR.

AwkwardPeak7 commented on src/all/asmhentai/src/eu/kanade/tachiyomi/extension/all/asmhentai/AsmHentai.kt:
```diff
@@ -1,274 +1,143 @@
-    override fun latestUpdatesNextPageSelector(): String? {
-        throw UnsupportedOperationException()
-    }
+    private val shortenTitleRegex = Regex("""(\[[^]]*]|[({][^)}]*[)}])""")
```
how about adding shorten title prefernece to the base class

cuong-tran replied:
HentaiFox doesn't seem to need this shortenTitle. Is it OK to put it in base class?

AwkwardPeak7 replied:
sure why not, you are gonna add other sites to this multisrc which do need it.

cuong-tran committed:
Move shortTitle to base clash

cuong-tran committed:
HentaiFox: add language keyword to search query

cuong-tran committed:
if all mangas in current searching page is of other language then include at least 1 entry so it can request for next page

cuong-tran committed:
Alternative methods for images parsing
Revert "remove images parsing setting, pick a default one"

This reverts commit e49e3aaeb74e3643abc2e303924da18a52491793.

# Conflicts:
#	lib-multisrc/galleryadults/src/eu/kanade/tachiyomi/multisrc/galleryadults/GalleryAdults.kt
#	src/all/asmhentai/src/eu/kanade/tachiyomi/extension/all/asmhentai/AsmHentai.kt
#	src/all/imhentai/src/eu/kanade/tachiyomi/extension/all/imhentai/IMHentai.kt

cuong-tran committed:
Fall back if failed to decode JSON

cuong-tran committed:
remove supportLatest from base class

AwkwardPeak7 commented on lib-multisrc/galleryadults/src/eu/kanade/tachiyomi/multisrc/galleryadults/GalleryAdults.kt:
```diff
@@ -0,0 +1,785 @@
+            setDefaultValue(false)
+        }.also(screen::addPreference)
+
+        SwitchPreferenceCompat(screen.context).apply {
```
I still don't think this should be an option to the user. keep the current fallback parsing logic but make this a class variable, to be overriden if we receive report of missing pages/unable load etc.

cuong-tran replied:
sure, that works too.

AwkwardPeak7 commented on lib-multisrc/galleryadults/src/eu/kanade/tachiyomi/multisrc/galleryadults/GalleryAdults.kt:
```diff
@@ -0,0 +1,785 @@
+        // Advance search
+        val advancedSearchFilters = filters.filterIsInstance<Filter.Text>()
+
+        return when {
```
consider splitting these differnet types into different functions, so they can be overriden specifically instead of the whole method

cuong-tran committed:
Remove preference for parsing page by page, switch to override val instead.

cuong-tran committed:
Split searchRequest into parts

cuong-tran committed:
Don't using generic Filter.Text to avoid other kind of text field which extensions might have
